### PR TITLE
Provide datum, index, and dataset to label formatter

### PR DESCRIPTION
**Pull request by dallinregehr**
An attempt at implementing #3088

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/plottable](https://github.com/palantir/plottable), @dallinregehr! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: src/core/formatters.ts [hunk 1]
```diff
@@ -1,7 +1,6 @@
 namespace Plottable {
 
-  export type Formatter = (d: any) => string;
-
+  export type Formatter = (d: any, datum?: any, index?: number, dataset?: Dataset) => string;
```
Comment (CalvinFernandez) [inline src/core/formatters.ts]: Nice! I'm a little concerned about the lack of type-safety afforded by including the parameters as optional arguments. 


Maybe we can include another `Formatter` type more tightly scoped to the components this change effects, namely `Pie` and `Bar` charts? Consider creating a new formatter type declared under the namespace `Plottable.Plots` called `LabelFormatter` or something which is of type `(d: any, datum: any, index: number, dataset: Dataset) => string;`. Then if you change the type of `_labelFormatter` to `LabelFormatter` in Pie and Bar, we can guarantee that datum, index and dataset will be passed to the labelFormatter callback for those charts while making no promises about the other `Formatter` instances.

Comment (dallinregehr) [reply src/core/formatters.ts]: Good idea. I'll update shortly.

File: src/plots/barPlot.ts [hunk 2]
```diff
@@ -212,17 +212,38 @@ namespace Plottable.Plots {
...
+    /**
+     * Sets the Formatter for the chart axis.
      *
      * @param {Formatter} formatter
      * @returns {Bar} The calling Bar Plot.
      */
-    public labelFormatter(formatter: Formatter): this;
-    public labelFormatter(formatter?: Formatter): any {
+    public axisLabelFormatter(formatter: Formatter): this;
+    public axisLabelFormatter(formatter?: Formatter): any {
```
Comment (CalvinFernandez) [inline src/plots/barPlot.ts:247]: Ah we can't do this its an api-breaking change

**Review by CalvinFernandez — CHANGES_REQUESTED** [review]
Hmm unfortunately we can't add an `axisLabelFormatter` endpoint and remove support for axis label formatting for those who've only defined a `labelFormatter`. 

The formatter call in stacked bar chart ( the thing thats messing up your current api ) is formatting the sum label that appears above the stack.
```
     9 < --- This label 
   [ 2 ]
   [ 3 ]
   [ 4 ]
```
One thing you could do to get around the api issue is to change `LabelFormatter` to `(d: any, datum: any[], index: number[], dataset: Dataset)` 


And then the stacked bar call would pass the formatting function the array of data points in the stack along with their indices. 


What are you thoughts on this?

**Comment by adidahiya** [comment]
Ping on above questions @dallinregehr ^

Also, heads up, there are now a bunch of conflicts because we have migrated the codebase to use modules rather than namespaces.

**Comment by dallinregehr** [comment]
Decided to open new PR (#3226) branched on migrated codebase to avoid messy conflict resolutions.

**Pull request closed without merging**